integration-tests/deployment/multiclient.go

```diff
@@ -39,53 +44,95 @@ type RPC struct {
 var _ OnchainClient = &MultiClient{}
 
 type MultiClient struct {
+	logger logger.Logger
 	*ethclient.Client
 	Backups     []*ethclient.Client
 	RetryConfig RetryConfig
 }
 
-func NewMultiClient(rpcs []RPC, opts ...func(client *MultiClient)) (*MultiClient, error) {
+func WithRetryConfig(attempts uint64, delay time.Duration) func(client *MultiClient) {
+	return func(client *MultiClient) {
+		client.RetryConfig = RetryConfig{
+			Attempts: attempts,
+			Delay:    delay,
+		}
+	}
+}
+
+func NewMultiClient(lggr logger.Logger, rpcs []RPC, opts ...func(client *MultiClient)) (*MultiClient, error) {
 	if len(rpcs) == 0 {
-		return nil, fmt.Errorf("No RPCs provided, need at least one")
+		return nil, fmt.Errorf("no RPCs provided, need at least one")
+	}
+	mc := &MultiClient{
+		logger: lggr,
 	}
-	var mc MultiClient
 	clients := make([]*ethclient.Client, 0, len(rpcs))
-	for _, rpc := range rpcs {
+	for i, rpc := range rpcs {
 		client, err := ethclient.Dial(rpc.WSURL)
 		if err != nil {
-			return nil, errors.Wrapf(err, "failed to dial %s", rpc.WSURL)
+			lggr.Warnf("failed to dial rpc %d ending in %s, moving to next one", i+1, rpc.WSURL[len(rpc.WSURL)-4:])
+			continue
 		}
 		clients = append(clients, client)
 	}
+	if len(clients) == 0 {
+		return nil, fmt.Errorf("all RPCs failed, try again with different URLs")
+	}
 	mc.Client = clients[0]
-	mc.Backups = clients[1:]
+	if len(clients) > 1 {
+		mc.Backups = clients[1:]
+	} else {
+		lggr.Warn("Only one RPC provided, no backups available")
+	}
 	mc.RetryConfig = defaultRetryConfig()
+	mc.logger = lggr
 
 	for _, opt := range opts {
-		opt(&mc)
+		opt(mc)
 	}
-	return &mc, nil
+	return mc, nil
+}
+
+func (mc *MultiClient) HeaderByNumber(ctx context.Context, number *big.Int) (*types.Header, error) {
+	var header *types.Header
+	err := mc.retryWithBackups(ctx, func(client *ethclient.Client) error {
+		var err error
+		header, err = client.HeaderByNumber(ctx, number)
+		return err
+	})
+	return header, err
+}
+
+func (mc *MultiClient) CallContract(ctx context.Context, call ethereum.CallMsg, blockNumber *big.Int) ([]byte, error) {
+	var result []byte
+	err := mc.retryWithBackups(ctx, func(client *ethclient.Client) error {
+		var err error
+		result, err = client.CallContract(ctx, call, blockNumber)
+		return err
+	})
+	return result, err
 }
 
 func (mc *MultiClient) TransactionReceipt(ctx context.Context, txHash common.Hash) (*types.Receipt, error) {
 	var receipt *types.Receipt
-	err := mc.retryWithBackups(func(client *ethclient.Client) error {
+	// TransactionReceipt might return ethereum.NotFound error if the transaction is not yet mined
+	err := mc.retryWithBackups(ctx, func(client *ethclient.Client) error {
 		var err error
 		receipt, err = client.TransactionReceipt(ctx, txHash)
 		return err
-	})
+	}, ethereum.NotFound)
```

Comment:
@connorwstein wdyt of this? I faced this during waiting for a transaction to get confirmed. Before it's confirmed it returns not found and back up logic tries to change the client. That's why added this to acceptable error so that it does not attempt to change client or retry unnecessarily if this error is encountered